### PR TITLE
Rewrite D3

### PR DESCRIPTION
**Pull request by jaimerson**
Line charts and pie charts working. The API is as follows:
```js
renderer = new Kiddo.Renderer('#chartAnchor');
renderer.lineChart(y_axis_title, json_data); // or renderer.pieChart(title, json_data)
```
TODO:
- [x] labels;
- [x] multi-series;
- [x] information on hover;
- [x] zooming (kinda; not yet refreshing the graph without leaving the page);
- [x] better styles;
- [x] pie charts;
- [x] integration with the actual API;
- [x] responsive;
- [x] bring back date controls;

----------------
image

### CODE REVIEW

**Comment by jaimerson** [comment]
image

**Comment by jaimerson** [comment]
Responsive test: image

**Comment by sbrossie** [comment]
Notes from the sync-up meeting to discuss what's next:

* Legend. We need something similar to [dataaddict web site](http://dataaddict.fr/prenoms/#basile,leon)
  * Legend box with x and y coordinates (moves on the other side of the graph when cursor moves)
  * Each line has its own legend

* Data points. Unfortunately our seed data is not very good, which makes it difficult to make a good demo (we are aware of it and we need to work on that), so in the meantime i am still wondering if having a switch in the code where we could hardcode json data for various graphs would not make sense. Would also make it easier for us to try your latest code without relying on existing KB + data running somewhere.

* Date and zooming functionality
  * We need one box where we can select the end date (by default this is today)
  * We need 4 little boxes below the graph (similar to dataadict site) where we can zoom-in zoom-out for last month, last 3 months, last 6 months, last year with regard to specific end date entered.

* Additional TODO to reintroduce the existing functionality we had before:
  * smoothing function, ..
  * Add ability to download raw data
  * Spinner (until graph loads)

**Comment by jaimerson** [comment]
kanaui

**Comment by jaimerson** [comment]
I think that's all. The date selection does not work with fake data, because it's fake data after all, and always returns the same json.

**Comment by sbrossie** [comment]
Items from the meeting:
* Legend
  * Better looking
  * Move away from the points && Legend overwrites graph
* Fix Standalone

When this is done sync up with Pierre, who will load some data on the cloud and iterate from there for final integration.

**Comment by jaimerson** [comment]
Standalone css fixed o/

**Comment by pierre** [comment]
> Standalone css fixed o/

It's a bit better indeed, but the outside container seems to have disappeared (everything now spans the entire width) and the top header is gone.

Also, the line graph doesn't seem to work.

Input data:

```
[{"type":"TIMELINE","title":"New accounts","data":[{"name":"count: 41","values":[{"x":"2015-06-01","y":0.0},{"x":"2015-06-02","y":3.0}]}]}]
```

Error:

```
Error: Invalid value for <path> attribute d="MNaN,430LNaN,0"
u @ d3.min.js:1
(anonymous function) @ d3.min.js:3
Y @ d3.min.js:1
_a.each @ d3.min.js:3
_a.attr @ d3.min.js:3
(anonymous function) @ kanaui.js:15491
Kiddo.LineChart.render @ kanaui.js:15479
Kiddo.Renderer.lineChart @ kanaui.js:15658
$.ready.d3.json.render @ kanaui.js:15716
(anonymous function) @ kanaui.js:15725
(anonymous function) @ d3.min.js:1

Error: Invalid value for <text> attribute transform="translate(NaN,0)"
u @ d3.min.js:1
(anonymous function) @ d3.min.js:3
Y @ d3.min.js:1
_a.each @ d3.min.js:
3_a.attr @ d3.min.js:3
(anonymous function) @ kanaui.js:15497
Kiddo.LineChart.render @ kanaui.js:15479
Kiddo.Renderer.lineChart @ kanaui.js:15658
$.ready.d3.json.render @ kanaui.js:15716
(anonymous function) @ kanaui.js:15725
(anonymous function) @ d3.min.js:1
t @ d3.min.js:1
u @ d3.min.js:1
```

**Comment by jaimerson** [comment]
kanaui

**Comment by jaimerson** [comment]
Also, @pierre, there was indeed a bug when there was only one line to plot, now it's fixed.

**Comment by pierre** [comment]
Thanks! I was able to render the graphs now.

Here are a few issues:

## Per-line legend overlapping

See the bottom right. But because we now have the dynamic box, we could get rid of it.

screen shot 2015-09-08 at 5 21 55 am

## X-axis legend overlapping

See after Sat 29.

screen shot 2015-09-08 at 5 25 45 am

## Match legend with lines

See previous image. It's hard to follow on the dynamic box which line matches which legend (maybe add a trailing colored dot, similar to http://dataaddict.fr/prenoms/#basile,leon).

## Remove extra blank space at the bottom of the dynamic box

The dynamic box seems to have an extra margin at the bottom.

## X-axis dots don't match data point

The highlighted point is the 18th, but it looks like it's aligned with the 20th on the X-axis. Also, the report started on 08/08, but the first point shown is 09/08.

screen shot 2015-09-08 at 5 30 46 am

## Pie chart legend overlapping

screen shot 2015-09-08 at 5 24 31 am

**Comment by jaimerson** [comment]
- Per-line legend overlapping: I think the best way to solve this is not to have this legend at all, and instead have the matching colors on the box (the trailing colored dot);
- X-axis legend overlapping: I'm not sure how to proceed on this, since the number of points in x-axis is unknown;
- Match legend with lines: See the first point;
- Remove extra blank space at the bottom of the dynamic box: I couldn't reproduce this, will try again later;
- X-axis dots don't match data point: I'll have a look on it;
- Pie chart legend overlapping: Same as per-line legend overlapping, it's probably best to show the labels separately.

**Comment by pierre** [comment]
> - *Per-line legend overlapping*: I think the best way to solve this is not to have this legend at all, and instead have the matching colors on the box (the trailing colored dot);

:+1: 

> - *X-axis legend overlapping*: I'm not sure how to proceed on this, since the number of points in x-axis is unknown;

Can we configure [d3.time](https://github.com/mbostock/d3/wiki/Time-Formatting) to return only Mondays for the day granularity (looks like Month and Year granularity is already correctly handled)? We would have only ticks for one day of the week on the x-axis, but moving the mouse over the graph would give us day-by-day details.

> - *Match legend with lines*: See the first point;

:+1: 

> - *Remove extra blank space at the bottom of the dynamic box*: I couldn't reproduce this, will try again later;

In my testing, the more lines are drawn on the graph, the bigger the blank space is at the bottom.

> - *Pie chart legend overlapping*: Same as per-line legend overlapping, it's probably best to show the labels separately.

:+1: (and we can use the same color dot trick to match sectors with legend entries)

**Comment by jaimerson** [comment]
kanaui-new

**Comment by jaimerson** [comment]
image
image
"Fixed" the x-axis overlapping issue by lowering the number of ticks shown, let me know if it's too lame (or if it does not work for all cases).

**Comment by jaimerson** [comment]
Only one left is pie chart legend overlapping.

**Comment by pierre** [comment]
> Per-line legend overlapping

Verified.

screen shot 2015-09-13 at 1 55 17 am

> X-axis legend overlapping

Verified. The last month option shows one tick per week, which I think is good.

screen shot 2015-09-13 at 1 58 00 am

> Match legend with lines

Verified.

screen shot 2015-09-13 at 2 12 43 am

> Remove extra blank space at the bottom of the dynamic box

Verified.

> X-axis dots don't match data point

Verified.

Besides the pie chat, we're looking really good for this PR! Once you have a fix, we can merge it and fix killbill-admin-ui-standalone (do you know if there are any changes required?).

FYI, here are a few enhancements / nice-to-haves that I noticed, but we can tackle them in a separate PR:

* Discreet support lines on the X and Y axis would be helpful (e.g. 5 equally spaced parallel to the X-axis, one for each X tick parallel to the Y-axis)
* Improve CSS of the legend box: date is not quite centered and the box is often a bit too large
* The overall standalone Kaui CSS could be improved: margins around the border of the screen, filter dates box, how we display the list of available reports
* A "no data" label would be nice if the graph has no data
* On the Y-axis, hide non-round ticks (e.g. show 1, 2 and 3, not 1.0, 1.2, 1.4, 1.6, etc.) as they often don't make sense (e.g. # of cancellations)

**Comment by jaimerson** [comment]
For the changes required in standalone, see https://github.com/killbill/killbill-admin-ui-standalone/pull/8.
I'll see what I can do for the rest when I have the time.

**Comment by sbrossie** [comment]
@jaimerson do you have any status (we 'd like to wrap up that project as soon as possible). It seems like we are getting close, the main thing on our end is the list from Pierre above.

**Comment by jaimerson** [comment]
Sorry for the delay, I think I will have time to finish those issues on the weekend.

**Comment by jaimerson** [comment]
image

**Comment by jaimerson** [comment]
image

**Comment by jaimerson** [comment]
Hope this settles it, let me know if there are any remaining issue.

**Comment by sbrossie** [comment]
Looking good.

At this point we probably don't want the fake data to load. Could you just add a switch in the code not not load it by default? And then i guess we are ready to merge...

**Comment by jaimerson** [comment]
It currently loads if the Rails env is set to development. But I will add a switch in the code.
